### ISSUE TITLE
Discussion Episode 8: Damn Functional Programming with Paco Estevez

### ISSUE THREAD

**Issue by sockeqwe** [artem-zinnatullin/TheContext-Podcast#57]

**Comment by artem-zinnatullin** [comment]
cc @pakoito

**Comment by pakoito** [comment]
![](http://i.imgur.com/z2ORn.gif)

**Comment by artem-zinnatullin** [comment]
Yup (we have custom `ofType()` which is filter + map under the hood, that what I was talking about), btw, @pakoito pls just write separate comments instead of editing one because nooooone receives emails about edited comments 😸

**Comment by artem-zinnatullin** [comment]
Btw, here is the code of problematic `ofType` (not `filter`, my bad):

```kotlin
// This can be applied to any observable no matter which type is emitted by original one 
// and may lead to no events in the stream if incompatible types are used.
inline fun <reified R : Any> Observable<*>.ofType1(): Observable<R> = ofType(R::class.java)

// This version has type bounds, but for some reason Kotlin compiler requires 
// you to also pass first type, though I think it should be able to infer it from the upstream.
inline fun <reified T, reified R : T> Observable<T>.ofType2(): Observable<R> 
= filter { it is R }.map { it as R }
```

Same problem as happens with `ofType1()` will occur if you'll use stardard RxJava's `ofType(Class)` in both 1.x and 2.x.

**Comment by artem-zinnatullin** [comment]
Examples:

```kotlin
Observable.just(1).ofType(String::class.java).subscribe { /* Compiles :( */ }

Observable.just(1).ofType1<String>().subscribe { /* Compiles :( */ }

Observable.just(1).ofType2<Int, String>().subscribe { /* Not compiles! But requires specifying upstream type */ }
```

**Comment by sebaslogen** [comment]
Fantastic program, I quite enjoyed it 👍
My suggestion would have been to split it into two shorter parts (there was a good separation of topics about half way through the program), it took me a while to find the free time to listen to the whole episode.

Thanks for explaining Monads, the concept is still exploding in my mind but your explanation is close to what I found in [this monads ELI5](https://www.reddit.com/r/AskProgramming/comments/54wzhf/eli5_monads_functional_programming/?st=j0f4w0cj&sh=fcd447c3) and the concept is starting to crystalize in my mind.

I really liked the part where you guys discuss the similarities between Mosby3 and Paco's solution, although I would like a deeper dive in the differences of each solution to understand the nuances. Super interesting topic 👌

Note: The testing robots thing is indeed an abstraction from the UI layer for testing and it's originally called PageObject pattern coming from the old wild web world. Martin Fowler introduced it here: https://martinfowler.com/bliki/PageObject.html

Keep up with the high quality of the topics! 💯
